### PR TITLE
Scrutinize R/U/GHF for compile time correct types

### LINKED ISSUE #710: Scrutinize R/U/GHF and improve its APIs
**Describe the current state of a part of the code**
In #688, we've done a lot of compile-time corrections to the code. One of the byproducts of this is that possibly the APIs that are used in the R/U/GHF modules (both in `QCMethod` as in `QCModel`) don't use the most up-to-date R/U/G types.

**Describe a change that would improve the code base**
It would be good if the HF modules would be scrutinized and checked if their APIs need to be updated to reflect the most up-to-date types we have.

**Additional context**
@xdvriend Could you take the lead for this? Of course, this would only be possible after #707 is closed.

### PR DESCRIPTION
**Pull request by xdvriend**
**Short description**
This PR looks to improve the API"'s of all three HF methods by using the compile-time correct types.

**Related issues**
closes #710 

**Additional context**

- [x] UTransformationMatrix in UHF

- [x] Rewrite RHF with operators

- [x] Rewrite GHF with operators

**Don't forget:**
- [x] if new files are created: update the Doxygen file, the collective gqcp.hpp header and the CMake files

### CODE REVIEW

**Comment by xdvriend** [comment]
After checking the RHF and GHF methods, I believe both already use their respective compile-time correct types.
I can however rewrite both methods with their respective one-electron operators, instead of a `SquareMatrix` for the `Fock Matrices`. This would be more in line with the way UHF is written.
I would love to hear your thoughts on this @lelemmen !

**Comment by lelemmen** [comment]
> After checking the RHF and GHF methods, I believe both already use their respective compile-time correct types.
> I can however rewrite both methods with their respective one-electron operators, instead of a `SquareMatrix` for the `Fock Matrices`. This would be more in line with the way UHF is written.
> I would love to hear your thoughts on this @lelemmen !

Yes, that's exactly the refactor I was looking for: refactoring the whole R/U/G HF module to work with operators instead of the raw matrices. Since you're the one working on it: feel free to adjust the estimate if you want to.

**Comment by xdvriend** [comment]
I believe I've adjusted everything to the corresponding compile-time correct types.

- Fock matrices are represented by their respective One-electron operators.
- Coefficient matrices are represented by their respective transformation matrices.
- The overlap matrix is represented by an operator. In UHF I used an `RSQOneElectronOperator`, since I didn't see the point using an unrestricted operator here, as `S` is always the same.

File: gqcp/include/QCMethod/HF/UHF/UHFSCFEnvironment.hpp [hunk 1]
```diff
@@ -55,10 +56,9 @@ class UHFSCFEnvironment {
     std::deque<VectorX<double>> orbital_energies_alpha;
     std::deque<VectorX<double>> orbital_energies_beta;
 
-    SquareMatrix<Scalar> S;  // the overlap matrix (of the scalar (AO) basis)
+    ScalarRSQOneElectronOperator<Scalar> S;  // the overlap matrix (of the scalar (AO) basis)
```
Comment (lelemmen) [inline gqcp/include/QCMethod/HF/UHF/UHFSCFEnvironment.hpp:59]:
```suggestion
    ScalarUSQOneElectronOperatorComponent<Scalar> S;  // the overlap matrix (of the scalar (AO) basis)
```

I think this fits better here.

File: gqcp/include/QCModel/HF/UHF.hpp [hunk 2]
```diff
@@ -49,8 +50,7 @@ class UHF {
     VectorX<double> orbital_energies_alpha;  // sorted by ascending energy
     VectorX<double> orbital_energies_beta;   // sorted by ascending energy
 
-    TransformationMatrix<Scalar> C_alpha;  // the coefficient matrix that expresses every alpha spatial orbital (as a column) in its underlying scalar basis
-    TransformationMatrix<Scalar> C_beta;   // the coefficient matrix that expresses every beta spatial orbital (as a column) in its underlying scalar basis
+    UTransformationMatrix<Scalar> C;  // the coefficient matrix that expresses every alpha spatial orbital (as a column) in its underlying scalar basis
```
Comment (lelemmen) [inline gqcp/include/QCModel/HF/UHF.hpp:53]:
```suggestion
    UTransformationMatrix<Scalar> C;  // The transformation between the current unrestricted spin-orbitals and the AOs.
```

File: gqcp/include/QCModel/HF/UHF.hpp [hunk 3]
```diff
@@ -68,17 +68,16 @@ class UHF {
      *  @param C_alpha                                  the coefficient matrix that expresses every alpha spatial orbital (as a column) in its underlying scalar basis
      *  @param C_beta                                   the coefficient matrix that expresses every beta spatial orbital (as a column) in its underlying scalar basis
      */
-    UHF(const size_t N_alpha, const size_t N_beta, const VectorX<double>& orbital_energies_alpha, const VectorX<double>& orbital_energies_beta, const TransformationMatrix<Scalar>& C_alpha, const TransformationMatrix<Scalar>& C_beta) :
+    UHF(const size_t N_alpha, const size_t N_beta, const VectorX<double>& orbital_energies_alpha, const VectorX<double>& orbital_energies_beta, const UTransformationMatrixComponent<Scalar>& C_alpha, const UTransformationMatrixComponent<Scalar>& C_beta) :
```
Comment (lelemmen) [inline gqcp/include/QCModel/HF/UHF.hpp:71]:
```suggestion
    UHF(const size_t N_alpha, const size_t N_beta, const VectorX<double>& orbital_energies_alpha, const VectorX<double>& orbital_energies_beta, const UTransformationMatrix<Scalar>& C) :
```

Maybe encapsulate both components from the start?

File: gqcp/include/QCModel/HF/UHF.hpp [hunk 4]
```diff
@@ -108,7 +107,7 @@ class UHF {
      *  @param C_alpha                                  the coefficient matrix that expresses every alpha spatial orbital (as a column) in its underlying scalar basis
      *  @param C_beta                                   the coefficient matrix that expresses every beta spatial orbital (as a column) in its underlying scalar basis
      */
-    UHF(const size_t N_alpha, const size_t N_beta, const TransformationMatrix<Scalar>& C_alpha, const TransformationMatrix<Scalar>& C_beta) :
+    UHF(const size_t N_alpha, const size_t N_beta, const UTransformationMatrixComponent<Scalar>& C_alpha, const UTransformationMatrixComponent<Scalar>& C_beta) :
```
Comment (lelemmen) [inline gqcp/include/QCModel/HF/UHF.hpp:110]: Idem here.

File: gqcp/include/QCModel/HF/UHF.hpp [hunk 5]
```diff
@@ -352,16 +351,16 @@ class UHF {
      *
      *  @return the coefficient matrix that expresses the sigma spin-orbitals (as a column) in its underlying scalar basis
      */
-    const TransformationMatrix<Scalar>& coefficientMatrix(const Spin sigma) const {
+    const TransformationMatrix<Scalar> coefficientMatrix(const Spin sigma) const {
```
Comment (lelemmen) [inline gqcp/include/QCModel/HF/UHF.hpp:355]: I think you can re-implement this using `this->C.component(sigma)`.

File: gqcp/include/QCModel/HF/UHF.hpp [hunk 6]
```diff
@@ -407,12 +406,12 @@ class UHF {
 
         switch (sigma) {
```
Comment (lelemmen) [inline gqcp/include/QCModel/HF/UHF.hpp:407]: ... and this one using `this->coefficientMatrix(sigma).numberOfOrbitals()`.

**Review by lelemmen — APPROVED** [review]
Good to go after some minor adjustments.

**Pull request merged**